design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,160 @@
+* Kubelet’s CPUManager is set to static - full-pcpus-only.
+* Kubevirt CR's defaultRuntimeClass is set to a high performance runtimeClass.
+* VM is set with `dedicatedCPUPlacement` and `IsolateEmulatorThread` enabled.
+* VM requests an even number of CPUs.
+
+When enabling `IsolateEmulatorThread` on a VM with even CPU count, Kubevirt is requesting one extra CPU for the virt-launcher pod. This is an internal change that is not exposed to the user.
+After the pod is scheduled, the now odd number of CPUs will collide with the CPU-Manager policy of using only full physical cores, and cause the kubelet to [reject](https://github.com/kubernetes/kubernetes/pull/101432) this pod.
+For example, let’s consider this VM spec:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        cpu:
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+          sockets: 1
+          cores: 4
+          threads: 2
+```
+kubelet will emit following event:
+```
+SMT Alignment Error: requested 9 cpus not multiple cpus per core = 2
+```
+The following picture helps show how this is a problem
+<img src="smt-slignment-error-diagram.png">
+
+## Goals
+Fix [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103) on the Kubevirt scope.
+
+## Non  Goals
+Fix the actual scheduling of the virt-launcher to a node that will reject it.
+
+## Definition of Users
+This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.
+
+## User Stories
+A user should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.
+
+## Repos
+- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Bug fix design proposal
+The bug fix proposal introduces an alpha annotation `kubevirt.io/full-pcpus-only: true`:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    metadata:
+      annotations:
+        kubevirt.io/full-pcpus-only: true
+    spec:
+      domain:
+        cpu:
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+          sockets: 1
+          cores: 4
+          threads: 2
+```
+
+When the VM is configured with `isolateEmulatorThread: true` and the above annotation, the virt-controller will set the virt-launcher pod with an even amount of CPUs, meaning:
+
+| User requested CPUs               |Even (2,4,6,8,...)| odd (1,3,4,5,7,9,...) |
+|-----------------------------------|------------------|-----------------------|
+| Extra emulatorThread CPUs for pod |        +2        | +1 (as usual)         |
+
+Notes:
+1. In both scenarios the guest will get the same amount of CPUs it requested.
+2. For simplicity reasons, the housekeeping cgroup will contain only 1 CPU. However, this is open for discussion.
```

Comment:
If a dedicated CPU is already present on the virt-launcher pod, but not used - it is a waste, as no other code could run on it.

Reply:
true. But notice that virt-hanlder will have to check
- at the VM level: the existence of this annotations
- at the host level:  if this CPU has a sibling, and what it is.
It's a more complex, and possible has more implications than we think. so I'm leaving it fir debate.

@vladikr @iholder101 Can you share your thoughts ?

Reply:
I agree, in general...
The emulator thread is being pinned to a CPUSet [1]
And this CPUSet is being used to build the housekeeping cgroup [2].

I think it shouldn't be too hard to update [1] to place the 2 over-allocated CPUs in this set.

[1] https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go#L364
[2] https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm.go#L2828